File: src/stages/reconcile.js
import { logger } from "../utils/logger.js";
import { progressBar } from "../utils/progress.js";
import { readMetadata } from "../core/exiftool.js";
import { readSidecar } from "../core/sidecar.js";
import { handlers } from "../handlers/index.js";
import { extname } from "path";
import { parsePhotoTakenTime } from "../core/datetime.js";

export async function reconcile(manifest, options) {
  // Stage 1: filter keepers. We use optional chaining (?.) to be safe if dedupe object is missing.
  // If dedupe was skipped, action is 'pending' or undefined, so !== 'delete' is true.
  const keepers = manifest.filter((entry) => entry.dedupe?.action !== "delete");

  // It is unlikely to happen, but theoretically possible.
  if (keepers.length === 0) {
    return;
  }

  // Stage 2: collect a Set of all tags required by all registered handlers.
  const readTags = [...new Set(Object.values(handlers).flatMap((h) => h.readTags))];

  // Calculate total number of batches needed (round up for partial batches)
  const batchSize = 512;
  const batches = Math.ceil(keepers.length / batchSize);

  const bar = progressBar("⧗ Reconciling metadata |{bar}| {percentage}% | {value}/{total} batches");

  bar.start(batches, 0);

  // Stage 3: sliding window loop to process files in chunks to prevent OOM.
  for (let offset = 0; offset < keepers.length; offset += batchSize) {
    // Stage 3.1: slice the batch (up to batchSize items)
    const batch = keepers.slice(offset, offset + batchSize);

    // Stage 3.2: prepare paths for exiftool.
    const batchPaths = batch.map((entry) => entry.mediaPath);

    // Stage 3.3: contains a Map<Path, ExifObject>.
    const batchExifMap = await readMetadata(batchPaths, readTags);

    // Stage 3.4: process the batch in parallel using promises.
    const tasks = batch.map(async (entry) => {
      // Look up tags returned by exiftool.
      const rawDiskTags = batchExifMap.get(entry.mediaPath);

      try {
        entry.metadata = await resolve(entry.mediaPath, entry.json?.path, rawDiskTags);
      } catch (error) {
        if (options.strict) {
          throw error;
        }

        logger.error(error.message);
        entry.metadata = {
          status: "error",
          writeTags: [],
          dates: [],
        };
      }
    });

    await Promise.all(tasks);

    bar.increment();
  }

  bar.stop();
}

async function resolve(mediaPath, jsonPath, rawDiskTags) {
  // If exiftool has not reported on a file, then something went wrong.
  if (!rawDiskTags) {
    throw new Error(`ExifTool did not report on ${mediaPath}, skipping.`);
  }

  const extension = extname(mediaPath).toLowerCase();
  const handler = handlers[extension];

  // If there is no handler for a file type, pixelkasten will not be able to process it.
  if (!handler) {
    throw new Error(`Encountered file of unsupported type ${extension}, skipping.`);
  }

  // Normalize raw, file type specific tags to a common shape.
  const diskData = handler.parse(rawDiskTags);

  // The starting point for the metadata object. If no writes happen, the status is "noop".
  const metadata = { status: "noop", writeTags: [], dates: [...diskData.dates] };

  // Retrieve and parse sidecar data.
  const sidecarData = await readSidecar(jsonPath);

  // If there is no sidecar data, there is nothing to resolve.
  if (!sidecarData) {
    return metadata;
  }

  // If there is no primary timestamp on disk, we can embed the value from sidecar.
  if (!diskData.timestamp && sidecarData.timestamp) {
    const { iso, exif } = parsePhotoTakenTime(sidecarData.timestamp);

    // Ensure that new timestamp gets written back to the file.
    metadata.writeTags.push(...handler.timestamp(exif));

    // Make sure that timestamp is stored as the primary date
    metadata.dates.unshift(iso);
  }

  if (!diskData.geo && sidecarData.geo) {
    metadata.writeTags.push(...handler.geo(sidecarData.geo));
  }

  if (metadata.writeTags.length > 0) {
    metadata.status = "processed";
  }

  return metadata;
}

File: src/pixelkasten/stages/reconcile.py
"""
Metadata reconciliation -- compare disk EXIF with sidecar data.

Ported from packages/core/src/stages/reconcile.js. For each manifest entry,
reads the disk EXIF via exiftool and compares with sidecar data. Queues
writeTags for any metadata missing from disk.
"""

import os
from collections.abc import Callable

from pixelkasten.core.datetime import parse_photo_taken_time
from pixelkasten.core.exiftool import read_metadata
from pixelkasten.core.handlers import handlers
from pixelkasten.core.manifest import can_keep
from pixelkasten.core.sidecar import read_sidecar

BATCH_SIZE = 512


def reconcile(
    manifest: list[dict],
    options: dict | None = None,
    on_progress: Callable[[int], None] | None = None,
) -> None:
    """
    Compare disk EXIF with sidecar data, queue writeTags for missing metadata.

    Mutates manifest entries in-place by adding 'metadata' dict with:
        status: "noop" | "processed" | "skipped" | "error"
        writeTags: list[str] -- exiftool tag=value pairs
        dates: list[str] -- ISO 8601 timestamps
        reason: str -- (only for skipped/error)

    Processes entries in batches of 512 for efficient exiftool invocation.
    """
    options = options or {}

    keepers = [e for e in manifest if can_keep(e)]
    if not keepers:
        return

    # Collect all tags required by all handlers (deduplicated).
    all_tags = list({tag for h in handlers.values() for tag in h.read_tags})

    # Process in batches to prevent OOM with large manifests.
    for offset in range(0, len(keepers), BATCH_SIZE):
        batch = keepers[offset : offset + BATCH_SIZE]
        batch_paths = [e["mediaPath"] for e in batch]

        raw_metadata = read_metadata(batch_paths, all_tags)

        for entry in batch:
            raw_disk_tags = raw_metadata.get(entry["mediaPath"])
            json_path = None
            if entry.get("json"):
                json_path = entry["json"].get("path")

            try:
                entry["metadata"] = _resolve(
                    entry["mediaPath"], json_path, raw_disk_tags, options
                )
            except Exception as e:
                entry["metadata"] = {
                    "status": "error",
                    "reason": str(e),
                    "writeTags": [],
                    "dates": [],
                }

        if on_progress is not None:
            on_progress(min(offset + BATCH_SIZE, len(keepers)))


def _resolve(
    media_path: str,
    json_path: str | None,
    raw_disk_tags: dict | None,
    options: dict,
) -> dict:
    """Resolve metadata for a single entry."""
    ext = os.path.splitext(media_path)[1].lower()
    handler = handlers.get(ext)

    # If there is no handler for a file type, skip it.
    if not handler:
        return {
            "status": "skipped",
            "reason": f"No metadata handler for {ext}",
            "writeTags": [],
            "dates": [],
        }

    # If exiftool has not reported on a file, something went wrong.
    if raw_disk_tags is None:
        raise RuntimeError(f"ExifTool did not report on {media_path}")

    # Normalize raw, file-type-specific tags to a common shape.
    disk_data = handler.parse(raw_disk_tags)

    metadata: dict = {
        "status": "noop",
        "writeTags": [],
        "dates": list(disk_data.dates),
    }

    # Retrieve and parse sidecar data.
    sidecar_data = read_sidecar(json_path)

    # If there is no sidecar data, there is nothing to resolve.
    if not sidecar_data:
        return metadata

    # If there is no primary timestamp on disk, use the value from sidecar.
    if not disk_data.timestamp and sidecar_data.get("timestamp"):
        ptt = parse_photo_taken_time(sidecar_data["timestamp"])

        # Queue timestamp for writing if embedding is enabled.
        if not options.get("skip_embed"):
            metadata["writeTags"].extend(handler.timestamp(ptt["exif"]))

        # Make sure that timestamp is stored as the primary date (needed for rename).
        metadata["dates"].insert(0, ptt["iso"])

    # Queue geo data for writing if embedding is enabled.
    if not options.get("skip_embed") and not disk_data.geo and sidecar_data.get("geo"):
        metadata["writeTags"].extend(handler.geo(sidecar_data["geo"]))

    if metadata["writeTags"]:
        metadata["status"] = "processed"

    return metadata
